optax/_src/transform.py

```diff
@@ -869,6 +869,64 @@ def update_fn(updates, state, params=None):
   return base.GradientTransformation(init_fn, update_fn)
 
 
+class ScaleByNewtonState(NamedTuple):
+  """State holding the inverse of the sum of gradient squares to date."""
+
+  hessian_inv: base.Updates
+
+
+def sherman_morrison(a_inv, u, v):
+  den = 1.0 + (u @ a_inv) @ v
+  a_inv -= a_inv @ jnp.outer(u, v) @ a_inv / den
+  return a_inv
+
+
+def scale_by_newton(eps: float = 1e-7) -> base.GradientTransformation:
+  """Rescale updates by multiplying by the inverse of an approximation of the
+  hessian.
+
+  References
+  ----------
+  [^1]: [Hazan, E., Agarwal, A. and Kale, S., 2007. Logarithmic regret
+  algorithms for online convex optimization. Machine Learning, 69(2-3),
+  pp.169-192](https://www.cs.princeton.edu/~ehazan/papers/log-journal.pdf)
+
+  Args:
+    eps: A small floating point value to avoid zero denominator.
+
+  Returns:
+    An (init_fn, update_fn) tuple.
+  """
+
+  def init_fn(params):
+    hessian_inv = jax.tree_map(
+      lambda t: jnp.eye(len(t.flatten()), dtype=t.dtype) / eps, params
```

Comment:
I have never implemented these sherman-morrison updates so I don't know whether this is actually an issue, but from looking at the code, I'd be worried about catastrophic cancellation when starting with jnp.eye / eps which has diagonal entries of O(10^7) and doing lots of additions and subtractions in the first step.

Is this something you have thought about? Is this a problem here? Perhaps there is a way to take the limit eps -> 0 explicitly for the first step to avoid this?

Thanks a lot!

Reply:
You are absolutely right.  In fact, in practice, this parameter may not be small. It should be chosen with some apriori knowledge about the scale of parameters and gradients (e.g. by following the formulas in the article).
Perhaps we should not define a default value for this parameter to let the user choose it?

Reply:
Okay, yes, letting the user choose it is helpful - do you have an intuition for how generally applicable the default value would be? If it always has to be tuned, I'd remove the default entirely and force the user to choose it.

Also, is there an explanation (or a paper or reference implementation) we can give the user to help them choose `eps`? I don't think the paper presents experiments with this, does it?

Reply:
@eserie